### DISCUSSION TITLE
Max volume limit

### DISCUSSION THREAD

**Discussion by golonac** [music-assistant/support#3288] [Feature requests and ideas]
Hi! I'm using Music Assistant in Home Assistant to stream music to a receiver (Onkyo, DLNA). The MA volume slider controls the receiver's volume setting directly, but setting it below 30 is inaudible, and above 60 is dangerously loud. I've had users accidentally change the volume way too high with the slider, as they are accustomed to 0=not audible, 1=barely audible, 100=loud but not deadly. I didn't find any way to change the limits in MA or HA - HA Onkyo DLNA integration lets me set a maximum volume but the setting is not used by MA, and MA's volume normalization can be used to restrict the max volume, but then the receiver needs to be turned very high to hear anything, which is again dangerous when switching to another source. Apologies if this exists or doesn't belong in MA & thanks for the great app!

In summary: a per-output setting that changes what volumes 0 and 100 map to in the volume slider.

**Comment by Swedlund** [comment]
We need this! Same situation here!

**Comment by rjgrandy** [comment]
+1 for this feature, would love to set a max and min level to be able to use the whole volume range

**Comment by nebriv** [comment]
Also would love this feature!

**Comment by oleost** [comment]
Having same issue with a Samsung soundbar, volume is incredible loud on «low values» on bar. Needs to be able to set maximum output

**Comment by LauriKem** [comment]
This would be amazing - trying to adjust a single player volume on a mobile screen results in easy accidental maxing of a player volume and almost broken eardrums. With this I could also attempt to equalize volumes of different players by bringing the loud ones down.

**Reply by OzGav** [reply]
Accidentally changing the volume has been fixed in 2.7.0 beta 4

**Comment by keesfluitman** [comment]
Just happened to my son as well.

**Comment by ZDBioHazard** [comment]
I have the same issue with my Denon AVR via AirPlay where <5% is inaudible and >30% is "get a new HOA rule named after you", making adjusting the volume quite precarious.

After I nearly vaporized the living room while trying to adjust the stereo volume, I set up the following automation in HA to limit the volume if it gets set too loud:

```yaml
alias: Stereo Volume Limiter
mode: single
triggers:
  - alias: When stereo volume goes over the limit
    trigger: template
    value_template: |-
      {{ ((state_attr("media_player.denon_avr", "volume_level") | default(None)) or 1.0) >
         (states("input_number.stereo_volume_max") | float / 100.0) }}'
actions:
  - alias: Limit the stereo volume
    action: media_player.volume_set
    target:
      entity_id: media_player.denon_avr
    data:
      volume_level: |-
        {{ min(
          (state_attr("media_player.denon_avr", "volume_level") | default(None)) or 1.0,
          (states('input_number.stereo_volume_max') | float / 100.0)
        ) }}
```

This automation _usually_ kicks in and lowers the volume quickly enough to simply give everyone nearby a good scare at most.

It's not a great solution, but hopefully it can help protect people's ears, equipment, and relationships with their neighbors until a proper solution can hopefully be developed.

**Comment by DutchJaFO** [comment]
Sonos has this as a feature on their app for their speakers.
And it really helps make them more controllable as you get a more granular control over the volume when '0 - 100' is magically translated to '0 - 25%' on the hardware.

It would be nice if we could set it from within MA and have it available as a virtual feature for player providers that don't support it directly.
Future iteration could even link this to user profiles (which will be useful when your kids want to play their music ;)).

**Comment by distante** [comment]
I have a similar situation, my kids gave Chromecast speakers and anything above 35% is extremely loud. 

On the HA UI I have setup some sliders that can not go above a controlled percent, but on the MA UI I always have to try to find the right spot and sometimes I go too loud. 

Would be nice to add a limiter or a multipler maybe?

**Comment by mael-queau** [comment]
For anyone still looking for a solution, what I've come up with is using the "Target level for volume normalization" value in the player settings. I've got a Denon AVR-X2000 that was really hard to dial in to a comfortable level from within Music Assistant, as a volume of 30 was too low, but about 36 was starting to get uncomfortable.

EDIT : Sadly, this actually doesn't work. I imagine it's because I had to set a target of -30dB which incurred a significant correction, but it ended up making huge volume differences from one track to another, even within the same album, not only ruining the listening experience but also making me adjust the volume from track to track. So sadly, I had to go back to disabling volume normalization altogether, and I am now back to square one.

**Comment by oleost** [comment]


**Reply by OzGav** [reply]
Yeah the problem with that is scaling in `__final_volume_level `(the model layer) means the reported volume is incorrect — the player state says "80%" but the device is at 40%. This breaks:

- Home Assistant integrations that read volume_level and expect the real value
- Any automation that sets volume to a specific level
- Group volume math — the relative volume adjustment (volume_dif) is calculated from the displayed (fake) values, so the proportional changes become wrong

There is also no support for external volume changes and I dont think group children will work properly (but I havent actually tried it). Also in this thread min volumes were also requested. 

I will open a PR shortly addressing this request. Not sure how long it will take to get approved or if my approach is the best but we will see.

**Reply by oleost** [reply]
Actually it was my intention to scale it like this, guess people have different use cases.

My use case is my soundbar at 45% is already at 120dB. By setting this one at max 50% it scales much better with the rest of the equipment I have. And I want 50% to be the new 100% (for all intended purposes)

Not sure how deep it will go if you set a limit and dont scale it, like all devices etc, will you show that 0-100% is available or somehow show that the range is not available ? Does this play nice ?

Could probably have an check marked that enables/disables scaling.

You had right with the group volume math, this is fixed now.

Implemented minimum volume as well now,

EDIT: Links removed.

**Reply by OzGav** [reply]
I have already opened a PR, it clamps the volume control to within the specified range. For the best UX you need to make frontend changes as well.
On your point about scaling that isn't what happens in the backend as final volumes are passed an an `int`. So any fractional gain in volume step you think you are getting is actually cut off.

**Comment by oleost** [comment]


**Reply by evilpig** [reply]
Tested with my Samsung Q930D which anything between 20-100 is too much especially in my apartment (will be nicer in a house but been frustrating while playing in a group with my google speakers all over.

Works perfectly! Thanks. Curious how this can be made as a plugin or addon to Music Assistant rather than having to use the Dev Server?

**Reply by evilpig** [reply]
Guess I am now running into some issues. Was working fine at first but made a Spotify Connect player for my group so I can access it from the Spotify app. 

```
2026-04-12 16:19:47.696 WARNING (MainThread) [music_assistant.Spotify Connect - New Tawny] Failed to send volume command via Spotify Web API: 'Retry-After'
2026-04-12 16:19:47.704 WARNING (MainThread) [music_assistant.Spotify Connect - New Tawny] Failed to send volume command via Spotify Web API: 'Retry-After'
2026-04-12 16:19:48.031 WARNING (MainThread) [music_assistant.Spotify Connect - New Tawny] Failed to send volume command via Spotify Web API: 'Retry-After'
2026-04-12 16:19:48.031 ERROR (MainThread) [aiohttp.server] Error handling request from IP
Traceback (most recent call last):
  File "/app/venv/lib/python3.13/site-packages/music_assistant/controllers/players/helpers.py", line 129, in execute
    await fn(self, *args, **kwargs)
  File "/app/venv/lib/python3.13/site-packages/music_assistant/controllers/players/controller.py", line 694, in cmd_group_volume
    await self.set_group_volume(player, volume_level)
  File "/app/venv/lib/python3.13/site-packages/music_assistant/controllers/players/controller.py", line 1743, in set_group_volume
    await asyncio.gather(*coros)
  File "/app/venv/lib/python3.13/site-packages/music_assistant/controllers/players/controller.py", line 3005, in _handle_cmd_volume_set
    await plugin_source.on_volume(volume_level)
  File "/app/venv/lib/python3.13/site-packages/music_assistant/providers/spotify_connect/__init__.py", line 511, in _on_volume
    await self._spotify_provider._put_data(f"me/player/volume?volume_percent={volume}")
  File "/app/venv/lib/python3.13/site-packages/music_assistant/helpers/throttle_retry.py", line 124, in wrapper
    return await func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/venv/lib/python3.13/site-packages/music_assistant/providers/spotify/provider.py", line 1244, in _put_data
    backoff_time = int(response.headers["Retry-After"])
                       ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^
KeyError: 'Retry-After'
```

```
2026-04-12 16:20:44.970 WARNING (MainThread) [music_assistant.Spotify Connect - New Tawny] Failed to send volume command via Spotify Web API: 'Retry-After'
2026-04-12 16:20:44.970 WARNING (MainThread) [music_assistant.webserver] players/cmd/group_volume: 'Retry-After'
2026-04-12 16:20:44.975 WARNING (MainThread) [music_assistant.Spotify Connect - New Tawny] Failed to send volume command via Spotify Web API: 'Retry-After'
2026-04-12 16:20:44.980 WARNING (MainThread) [music_assistant.Spotify Connect - New Tawny] Failed to send volume command via Spotify Web API: 'Retry-After'
2026-04-12 16:21:03.959 INFO (MainThread) [music_assistant.Spotify Connect - Soundbar] Spotify Connect session disconnected
2026-04-12 16:21:11.595 INFO (MainThread) [music_assistant.Spotify Connect - Soundbar] Spotify Connect session connected for user: evilpigster
2026-04-12 16:21:15.283 INFO (MainThread) [music_assistant.Spotify Connect - Soundbar] Starting Spotify Connect playback [spotify_connect--RnQM3R2u] on player upc8a6eff767a5
2026-04-12 16:21:32.433 INFO (MainThread) [music_assistant.Spotify Connect - New Tawny] Spotify Connect session disconnected
2026-04-12 16:21:39.653 ERROR (MainThread) [music_assistant] Error doing task: Task exception was never retrieved
Traceback (most recent call last):
  File "/app/venv/lib/python3.13/site-packages/music_assistant/controllers/players/helpers.py", line 129, in execute
    await fn(self, *args, **kwargs)
  File "/app/venv/lib/python3.13/site-packages/music_assistant/controllers/players/controller.py", line 1031, in select_source
    raise PlayerCommandFailed(f"Player {player.state.name} is currently grouped")
music_assistant_models.errors.PlayerCommandFailed: Player Q-Series Soundbar is currently grouped

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/app/venv/lib/python3.13/site-packages/music_assistant/controllers/players/helpers.py", line 141, in wrapper
    await execute()
  File "/app/venv/lib/python3.13/site-packages/music_assistant/controllers/players/helpers.py", line 131, in execute
    raise PlayerCommandFailed(str(err)) from err
music_assistant_models.errors.PlayerCommandFailed: Player Q-Series Soundbar is currently grouped
2026-04-12 16:22:06.977 INFO (MainThread) [music_assistant.Spotify Connect - Auto] Spotify Connect session disconnected
2026-04-12 16:22:27.928 WARNING (MainThread) [music_assistant.Spotify Connect - Soundbar] Failed to send volume command via Spotify Web API: 404, message='Not Found', url='https://api.spotify.com/v1/me/player/volume?volume_percent=9'
2026-04-12 16:22:27.929 WARNING (MainThread) [music_assistant.webserver] players/cmd/group_volume: 404, message='Not Found', url='https://api.spotify.com/v1/me/player/volume?volume_percent=9'
2026-04-12 16:22:28.079 WARNING (MainThread) [music_assistant.Spotify Connect - Soundbar] Failed to send volume command via Spotify Web API: 404, message='Not Found', url='https://api.spotify.com/v1/me/player/volume?volume_percent=10'
2026-04-12 16:22:28.079 WARNING (MainThread) [music_assistant.webserver] players/cmd/group_volume: 404, message='Not Found', url='https://api.spotify.com/v1/me/player/volume?volume_percent=10'
2026-04-12 16:22:28.769 WARNING (MainThread) [music_assistant.Spotify Connect - Soundbar] Failed to send volume command via Spotify Web API: 404, message='Not Found', url='https://api.spotify.com/v1/me/player/volume?volume_percent=11'
2026-04-12 16:22:28.769 WARNING (MainThread) [music_assistant.webserver] players/cmd/group_volume: 404, message='Not Found', url='https://api.spotify.com/v1/me/player/volume?volume_percent=11'
2026-04-12 16:22:45.150 WARNING (MainThread) [music_assistant.Spotify Connect - Soundbar] Failed to send volume command via Spotify Web API: 404, message='Not Found', url='https://api.spotify.com/v1/me/player/volume?volume_percent=13'
2026-04-12 16:22:45.152 WARNING (MainThread) [music_assistant.webserver] players/cmd/group_volume: 404, message='Not Found', url='https://api.spotify.com/v1/me/player/volume?volume_percent=13'
2026-04-12 16:23:02.632 WARNING (MainThread) [music_assistant.Spotify Connect - Soundbar] Failed to send volume command via Spotify Web API: 404, message='Not Found', url='https://api.spotify.com/v1/me/player/volume?volume_percent=6'
2026-04-12 16:23:02.633 WARNING (MainThread) [music_assistant.webserver] players/cmd/group_volume: 404, message='Not Found', url='https://api.spotify.com/v1/me/player/volume?volume_percent=6'
2026-04-12 16:23:02.786 WARNING (MainThread) [music_assistant.Spotify Connect - Soundbar] Failed to send volume command via Spotify Web API: 404, message='Not Found', url='https://api.spotify.com/v1/me/player/volume?volume_percent=4'
2026-04-12 16:23:02.801 WARNING (MainThread) [music_assistant.webserver] players/cmd/group_volume: 404, message='Not Found', url='https://api.spotify.com/v1/me/player/volume?volume_percent=4'
```

**Reply by OzGav** [reply]
Sorry, this is not a support thread for other people’s forks. There is a PR open to properly support this.

**Reply by oleost** [reply]
> Sorry, this is not a support thread for other people’s forks. There is a PR open to properly support this.

Sorry for bringing that into her. Are your PR ready to be tested ? https://github.com/music-assistant/server/pull/3360

**Comment by OzGav** [comment]
This will be available in 2.9.0b8

**Comment by IntMainLoop** [comment]
A 'volume limit' feature may indeed be useful in its own right, but it doesn't solve the root problem experienced by most users, which is that the linear ui control element is not being correctly modified into a logarithmic control element as is required for audio volume controls to operate intuitively with regard to human hearing.

https://github.com/orgs/music-assistant/discussions/5423

**Comment by OzGav** [comment]
The feature request has been implemented successfully. As such this feature request will be closed and locked. If there are enhancements desired then open a new [feature request](https://github.com/orgs/music-assistant/discussions/categories/feature-requests-and-ideas). If there is a question about how this works then check the [documentation](https://beta.music-assistant.io/), or ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord. If there are problems then ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord or open an [issue](https://github.com/music-assistant/support/issues). Thanks for your interest in seeing this idea being brought to life.
